### PR TITLE
🚧 [HAP-1018][JENKINS-56073] Fixes JCasC Compatibility

### LINKED ISSUE #107: Support JCasC
Currently the plugin doesn't support [JCasC](https://jenkins.io/projects/jcasc/).

When an Artifactory server is configured in the global config, the JCasC export fails as follows
```
unclassified:
  artifactoryBuilder:
    artifactoryServers: "FAILED TO EXPORT org.jfrog.hudson.ArtifactoryBuilder$DescriptorImpl#artifactoryServers:\
      \ \nio.jenkins.plugins.casc.ConfiguratorException: Can't read attribute 'serverId'\
      \ from org.jfrog.hudson.ArtifactoryServer@12d71b12\n\tat io.jenkins.plugins.casc.Attribute._getValue(Attribute.java:252)\n\
      \tat io.jenkins.plugins.casc.Attribute.getValue(Attribute.java:174)\n\tat io.jenkins.plugins.casc.impl.configurators.DataBoundConfigurator.describe(DataBoundConfigurator.java:236)\n\
      \tat io.jenkins.plugins.casc.Attribute.describe(Attribute.java:192)\n\tat io.jenkins.plugins.casc.BaseConfigurator.compare(BaseConfigurator.java:387)\n\
      \tat io.jenkins.plugins.casc.impl.configurators.DescriptorConfigurator.describe(DescriptorConfigurator.java:61)\n\
      \tat io.jenkins.plugins.casc.impl.configurators.GlobalConfigurationCategoryConfigurator.describe(GlobalConfigurationCategoryConfigurator.java:105)\n\
      \tat io.jenkins.plugins.casc.impl.configurators.GlobalConfigurationCategoryConfigurator.lambda$describe$6(GlobalConfigurationCategoryConfigurator.java:98)\n\
      \tat java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)\n\
      \tat java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)\n\
      \tat java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)\n\
      \tat java.util.Iterator.forEachRemaining(Iterator.java:116)\n\tat java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)\n\
      \tat java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:481)\n\
      \tat java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)\n\
      \tat java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)\n\
      \tat java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)\n\
      \tat java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)\n\
      \tat java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)\n\
      \tat io.jenkins.plugins.casc.impl.configurators.GlobalConfigurationCategoryConfigurator.describe(GlobalConfigurationCategoryConfigurator.java:98)\n\
      \tat io.jenkins.plugins.casc.impl.configurators.GlobalConfigurationCategoryConfigurator.describe(GlobalConfigurationCategoryConfigurator.java:32)\n\
      \tat io.jenkins.plugins.casc.ConfigurationAsCode.export(ConfigurationAsCode.java:411)\n\
      \tat io.jenkins.plugins.casc.ConfigurationAsCode.doExport(ConfigurationAsCode.java:401)\n\
      \tat java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627)\n\
      \tat org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:343)\n\
      \tat org.kohsuke.stapler.interceptor.RequirePOST$Processor.invoke(RequirePOST.java:77)\n\
      \tat org.kohsuke.stapler.PreInvokeInterceptedFunction.invoke(PreInvokeInterceptedFunction.java:26)\n\
      \tat org.kohsuke.stapler.Function.bindAndInvoke(Function.java:184)\n\tat org.kohsuke.stapler.Function.bindAndInvokeAndServeResponse(Function.java:117)\n\
      \tat org.kohsuke.stapler.MetaClass$1.doDispatch(MetaClass.java:130)\n\tat org.kohsuke.stapler.NameBasedDispatcher.dispatch(NameBasedDispatcher.java:58)\n\
      \tat org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:739)\n\tat org.kohsuke.stapler.Stapler.invoke(Stapler.java:870)\n\
      \tat org.kohsuke.stapler.MetaClass$10.dispatch(MetaClass.java:384)\n\tat org.kohsuke.stapler.Stapler.tryInvoke(Stapler.java:739)\n\
      \tat org.kohsuke.stapler.Stapler.invoke(Stapler.java:870)\n\tat org.kohsuke.stapler.Stapler.invoke(Stapler.java:668)\n\
      \tat org.kohsuke.stapler.Stapler.service(Stapler.java:238)\n\tat javax.servlet.http.HttpServlet.service(HttpServlet.java:790)\n\
      \tat org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:841)\n\
      \tat org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1650)\n\
      \tat hudson.util.PluginServletFilter$1.doFilter(PluginServletFilter.java:154)\n\
      \tat hudson.util.PluginServletFilter.doFilter(PluginServletFilter.java:157)\n\
      \tat org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)\n\
      \tat hudson.security.csrf.CrumbFilter.doFilter(CrumbFilter.java:64)\n\tat org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)\n\
      \tat hudson.security.ChainedServletFilter$1.doFilter(ChainedServletFilter.java:84)\n\
      \tat hudson.security.ChainedServletFilter.doFilter(ChainedServletFilter.java:90)\n\
      \tat hudson.security.HudsonFilter.doFilter(HudsonFilter.java:171)\n\tat org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)\n\
      \tat org.kohsuke.stapler.compression.CompressionFilter.doFilter(CompressionFilter.java:49)\n\
      \tat org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)\n\
      \tat hudson.util.CharacterEncodingFilter.doFilter(CharacterEncodingFilter.java:82)\n\
      \tat org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)\n\
      \tat org.kohsuke.stapler.DiagnosticThreadNameFilter.doFilter(DiagnosticThreadNameFilter.java:30)\n\
      \tat org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1637)\n\
      \tat org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:533)\n\
      \tat org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:143)\n\
      \tat org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:524)\n\
      \tat org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)\n\
      \tat org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:190)\n\
      \tat org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:1595)\n\
      \tat org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:188)\n\
      \tat org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1253)\n\
      \tat org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:168)\n\
      \tat org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:473)\n\
      \tat org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:1564)\n\
      \tat org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:166)\n\
      \tat org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1155)\n\
      \tat org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)\n\
      \tat org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:219)\n\
      \tat org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:126)\n\
      \tat org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:132)\n\
      \tat org.eclipse.jetty.server.Server.handle(Server.java:564)\n\tat org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:317)\n\
      \tat org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:251)\n\
      \tat org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:279)\n\
      \tat org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:110)\n\tat\
      \ org.eclipse.jetty.io.ChannelEndPoint$2.run(ChannelEndPoint.java:124)\n\tat\
      \ org.eclipse.jetty.util.thread.Invocable.invokePreferred(Invocable.java:128)\n\
      \tat org.eclipse.jetty.util.thread.Invocable$InvocableExecutor.invoke(Invocable.java:222)\n\
      \tat org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:294)\n\
      \tat org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:199)\n\
      \tat org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:672)\n\
      \tat org.eclipse.jetty.util.thread.QueuedThreadPool$2.run(QueuedThreadPool.java:590)\n\
      \tat java.lang.Thread.run(Thread.java:748)\n"
```

### PR DESCRIPTION
**Pull request by jetersen**
[JENKINS-56073](https://issues.jenkins-ci.org/browse/JENKINS-56073)
[HAP-1018](https://www.jfrog.com/jira/browse/HAP-1018)

Not sure about the `pom.xml` but it WORKS, I can use `hpi:run` 🤣

made it describable time to make it JCasC compatible!

fixes #107 

Working in progress UI:
image

### CODE REVIEW

**Comment by casz** [comment]
@darxriggs feel free to take a look :)

File: pom.xml [hunk 1]
```diff
@@ -30,11 +30,9 @@
     <url>http://wiki.jenkins-ci.org/display/JENKINS/Artifactory+Plugin</url>
 
     <properties>
-        <!-- Minimal version required by Pipeline dependencies in 2.14.0 -->
-        <jenkins.version>1.642.3</jenkins.version>
+        <!-- Minimal version for Java 8 support -->
+        <jenkins.version>2.60.3</jenkins.version>
         <java.level>8</java.level>
-        <!-- TODO: enforcer is full of upper bound dependency issues, including real ones -->
-        <enforcer.skip>true</enforcer.skip>
```
Comment (casz) [inline pom.xml]: Yup that was evil to turn off enforcer 😭

**Comment by casz** [comment]
@eyalbe4 I plan to deprecate/migrate legacy username/password to credentials plugin. Much like I did for https://github.com/jenkinsci/violation-comments-to-stash-plugin/pull/45

File: pom.xml [hunk 2]
```diff
@@ -399,32 +397,134 @@
...
-            <scope>test</scope>
-        </dependency>
         <dependency>
             <groupId>org.jfrog.artifactory.client</groupId>
             <artifactId>artifactory-java-client-services</artifactId>
             <version>2.6.2</version>
             <scope>test</scope>
         </dependency>
         <!--/Tests-->
+        <!--- Upper bound dependencies -->
```
Comment (darxriggs) [inline pom.xml]: Shouldn't the upper bound restrictions not just be placed inside `<dependencyManagement>`?

File: src/main/resources/org/jfrog/hudson/ArtifactoryServer/config.jelly [hunk 3]
```diff
@@ -0,0 +1,30 @@
+<?jelly escape-by-default='true'?>
+
+<jelly:jelly xmlns:jelly="jelly:core" xmlns:f="/lib/form">
+  <f:entry field="id" title="Server ID" help="/plugin/artifactory/help/ArtifactoryBuilder/help-serverId.html">
```
Comment (darxriggs) [inline src/main/resources/org/jfrog/hudson/ArtifactoryServer/config.jelly]: I suggest to stick to the default naming convention for help files. Then the `help` attribute is not required anymore.

File: src/main/java/org/jfrog/hudson/ArtifactoryServer.java [hunk 4]
```diff
@@ -414,4 +430,35 @@ public ConverterImpl(XStream2 xstream) {
...
+        public static final int DEFAULT_CONNECTION_RETRY_NUMBER = 3;
+        public static final int DEFAULT_CONNECTION_TIMEOUT = 300;    // 5 Minutes
+
+        @Override
+        @NonNull
+        public String getDisplayName() {
+            return "Artifactory Server";
+        }
+
+        public ListBoxModel doFillConnectionRetryItems() {
```
Comment (darxriggs) [inline src/main/java/org/jfrog/hudson/ArtifactoryServer.java]: Or just use `<f:number clazz="positive-number" min="0" step="1" />` in the Jelly file.

File: src/main/resources/org/jfrog/hudson/ArtifactoryServer/config.jelly (hunk 3 above)
Comment (casz) [reply src/main/resources/org/jfrog/hudson/ArtifactoryServer/config.jelly]: Copy paste for now 😅

File: src/main/java/org/jfrog/hudson/ArtifactoryServer.java (hunk 4 above)
Comment (casz) [reply src/main/java/org/jfrog/hudson/ArtifactoryServer.java]: Fixed

File: src/main/resources/org/jfrog/hudson/ArtifactoryServer/config.jelly (hunk 3 above)
Comment (casz) [reply src/main/resources/org/jfrog/hudson/ArtifactoryServer/config.jelly]: Fixed, though could add some more help files

**Comment by eyalbe4** [comment]
@casz,
Thank you so much for contributing this!
Questions - 
You wrote that you're planning to remove the username/password support. Are existing jobs going to be automatically migrated to be using the Credentials Plugin?
Will people still have the option of using username/password? The reason why we allowed to choose between the two options is because we were concerned that some organisations will not allow using the Credentials Plugin for storing Artifactory credentials.
What is your opinion about this? Others are of course invited to join this discussion

**Comment by casz** [comment]
jobs would be migrated.
Why would someone have something against Credentials Plugin?

**Comment by darxriggs** [comment]
@eyalbe4 The Credentials Plugin provides more functionality than simply storing a password as `Secret`; e.g. multiple storage providers, auditing, ... Therefore plugins should use it. An extensive documentation can be found at https://github.com/jenkinsci/credentials-plugin/tree/master/docs.

**Comment by eyalbe4** [comment]
Thanks @casz and @darxriggs - 
Actually, if we're removing the username/passwords fields only in the *Configure Syetem* area and not in the jobs themselves, then I guess this is okay.
My only concern was for the credentials configured when editing the jobs, because there might be some users who wouldn't want to allow choosing a user (without actually typing the password) when configuring a job.

**Comment by casz** [comment]
Even on job with the right permissions they should still be able to configure with credentials. As an example on folder level you can give them credentials permission.

**Comment by eyalbe4** [comment]
OK @casz.
We can discuss this further if this becomes relevant.

**Comment by darxriggs** [comment]
Hi everybody, what's the plan to make some progress with this pull request?

**Comment by casz** [comment]
I got distracted 😓 There is a lot of work to fix.
But after hours working on it.
We could also make the easy fix and just add getters that matches the `@DataBoundConstructor` and call it a day.

Depends if @eyalbe4 likes to have some tests that confirm it working with JCasC

**Comment by tempora-mutantur** [comment]
@casz Hi.
May be we should split this PR, then?
Maybe I can help somehow?
Because JCasC became more mature, it would be nice to finish this PR.

**Comment by eyalbe4** [comment]
@casz,
Looks like I missed your comment from a month ago. I can help with some manual testing following your changes.

**Comment by darxriggs** [comment]
Just wanted to ping everybody so we could make some more progress.

@casz What exactly is still to be done for a basic JCasC compatibility?

@eyalbe4 did you already had the time to have a deeper look at it? I think @casz is waiting for some answer from your side.

**Comment by casz** [comment]
I can make a simple fix, but there is no telling without tests that it will be broken again. To add tests... This plugin needs to be refactored to modern times.

**Comment by casz** [comment]
@darxriggs I think I want to start the work in smaller PRs there is a lot of fixing needed to be done.

**Comment by eyalbe4** [comment]
@casz,
We're planning to start working on bumping the core Jenkins version used by the plugin. Hopefully this will make it easier for you  to add support for JCaasC.
I think smaller PRs will be easier to test and verify.

**Comment by darxriggs** [comment]
@eyalbe4 may I ask for the current status of the Jenkins version bumping?

**Comment by eyalbe4** [comment]
@darxriggs,
@liron-shalom created https://github.com/jfrog/jenkins-artifactory-plugin/pull/251 which bups the Jenkins core version. It even passed code review and we're planning to this during Match.
Awesome work @liron-shalom!

**Comment by eyalbe4** [comment]
@Jetersen and all - https://github.com/jfrog/jenkins-artifactory-plugin/pull/251 is released as part of version 3.6.0.
I believe this PR can be closed.

**Comment by sztomi** [comment]
@eyalbe4 it's a little unclear to me why this was closed and reopened. Does this mean the Artifactory plugin compatibility is fixed?

**Comment by eyalbe4** [comment]
@sztomi - sorry about that. 
This should stay open, because JCasC is still not supposed. The Jenkins core API has been undated though, which is a step in the right direction.

**Comment by imoazeem** [comment]
Any update regarding the compatibility of Artifactory plugin with JCasC

**Comment by peterloron** [comment]
Status update? This is causing serious issues for us.

**Comment by jetersen** [comment]
Doing smaller PRs, one of many to come.
https://github.com/jfrog/jenkins-artifactory-plugin/pull/285

**Comment by eyalbe4** [comment]
@Or-Gabay created https://github.com/jfrog/jenkins-artifactory-plugin/pull/298 in an attempt to support JCasC. We'll publish a snapshot version of the Artifactory Plugin with this code, so that everyone can evaluate it and confirm that it resolves the issue.

**Comment by timja** [comment]
> @Or-Gabay created #298 in an attempt to support JCasC. We'll publish a snapshot version of the Artifactory Plugin with this code, so that everyone can evaluate it and confirm that it resolves the issue.

Hi @eyalbe4 

Is there a reason there's no automated tests?
JCasC provides a mature testing framework to allow automated verification of the change and prevent regression,

See here:
https://github.com/jenkinsci/configuration-as-code-plugin/blob/master/docs/PLUGINS.md#how-to-test

**Comment by eyalbe4** [comment]
Thanks for sharing this @timja!
Adding a comment about it on #298.

**Comment by Or-Gabay** [comment]
I will get into automated tests very soon after I finish my current task.
Meanwhile, we just released Jenkins Artifactory plugin v3.7.1 which includes the JCasC [fix](https://github.com/jfrog/jenkins-artifactory-plugin/commit/1a6757915082783e2825248bd80f982aedc2aefc). I will appreciate your feedback!.

**Comment by Or-Gabay** [comment]
@timja, here is a [SNAPSHOT](https://oss.jfrog.org/artifactory/oss-snapshot-local/org/jenkins-ci/plugins/artifactory/3.7.x-SNAPSHOT/artifactory-3.7.x-20200723.060106-12.hpi) which includes the JCasC fix.
I will appreciate your feedback!.

**Comment by loganHighland7** [comment]
I saw the merge for this on the upstream got reverted, was there a specific reason? On the snapshot version I was still not able to see the jcasc for this plugin.

**Comment by eyalbe4** [comment]
@loganHighland7,
The code was reverted because of some issues it caused, but it wad then reimplemented and merged. 
Can you please elaborate about the issue you see with the above snapshot version? What are you not able to see? What is not working in the context of JCasC with this snapshot?

**Comment by loganHighland7** [comment]
@eyalbe4 
Manually installed the plugin on a Jenkins system, set up a test artifactory server in the config manually with the plugin through the global configuration menu. After, when attempting to load the jcasc through the ui, the section for this plugin still shows a long error message and I am not able to see the jcasc for what I just set.

```
FAILED TO EXPORT
      org.jfrog.hudson.ArtifactoryBuilder$DescriptorImpl#artifactoryServers: io.jenkins.plugins.casc.ConfiguratorException: Can't read attribute 'serverId' from org.jfrog.hudson.ArtifactoryServer@4040c36e
        at io.jenkins.plugins.casc.Attribute._getValue(Attribute.java:462)
        at io.jenkins.plugins.casc.Attribute.getValue(Attribute.java:236)
        at io.jenkins.plugins.casc.impl.configurators.DataBoundConfigurator.describe(DataBoundConfigurator.java:290)
        at io.jenkins.plugins.casc.Attribute._describe(Attribute.java:329)
        at io.jenkins.plugins.casc.Attribute.describe(Attribute.java:258)
        at io.jenkins.plugins.casc.Configurator.describe(Configurator.java:182)
        at io.jenkins.plugins.casc.impl.configurators.GlobalConfigurationCategoryConfigurator.describe(GlobalConfigurationCategoryConfigurator.java:115)
        at io.jenkins.plugins.casc.impl.configurators.GlobalConfigurationCategoryConfigurator.lambda$describe$3(GlobalConfigurationCategoryConfigurator.java:108)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
        at java.util.Iterator.forEachRemaining(Iterator.java:116)
        at java.util.Spliterators$IteratorSpliterator.forEachRemaining(Spliterators.java:1801)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
        at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
        at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:418)
        at io.jenkins.plugins.casc.impl.configurators.GlobalConfigurationCategoryConfigurator.describe(GlobalConfigurationCategoryConfigurator.java:108)
        at io.jenkins.plugins.casc.impl.configurators.GlobalConfigurationCategoryConfigurator.describe(GlobalConfigurationCategoryConfigurator.java:33)
```

**Comment by Or-Gabay** [comment]
@loganHighland7, looks like you are still using the old plugin, try to restart Jenkins after uninstalling the Artifactory plugin so no cache is left.

**Comment by loganHighland7** [comment]
After deleting plugin, restarting, and reinstalling it does appear to be working now. Thanks!

**Comment by eyalbe4** [comment]
Thanks for the feedback @loganHighland7!
All - if anyone else would like to provide some feedback for @Or-Gabay's snapshot, we'd greatly appreciate it!
We're aiming to release this by the end of this month.

**Comment by eyalbe4** [comment]
All - JCasC Compatibility is supported as part of the recent versions released. I'm therefore closing this one.

**Pull request closed without merging**